doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,412 @@
+  version data to enable interpretation by future downstream systems.
+- Prior art (e.g. OTIO) encodes versions in their data structure,
+  presumably for good reason.
+
+#### Cons
+- No "primary" version across the session, increasing implementation
+  complexity.
+- Introduces "Try and see" negotiation on the host side, although
+  introspection behaviour to negate this could be introduced at a later
+  date, similar to other introspection mechanisms. (See [In data -
+  Extended](#in-data---extended)).
+- Would likely require some utilities in `traitgen`, otherwise, due to
+  each new version effectively being a brand new trait, constructing the
+  correct trait view would be become a tedious dance.
+- `entityTraits` becomes more complex, as we either have to declare it
+  version agnostic somehow, or managers would need to return the correct
+  versions/entire swathes of versions. (See [What's
+  versioned](#whats-versioned)).
+- Unclear how specification versioning is to work without an out-of-data
+  version.
+- The most obvious implementation, (using `traitID` to store the
+  version), would effectively make any change to a trait version a
+  breaking change. Allowing non breaking changes (such as adding a new
+  property) would require significant rework to the `traitgen` output.
+
+### In data - Extended
+The implementation of in-data versioning is very basic, which is part of
+its strength, but it necessitates a "try and see" approach on the part
+of the host. If the host wants to be adaptive to the manager, it may
+benefit from a more up front introspection of the supported manager
+versions.
+
+To this end, the `supportedSchemaVersions` method (or similar concept,)
+from Option 2 could be orthogonally added without any conceptual
+conflict. This would have the added benefit of allowing hosts to also
+support a range of version/have backwards compatibility, which is a huge
+benefit.
+
+This stops short of negotiation, being merely introspection similar to
+our other manager introspection methods.
+
+### What's versioned?
+In this example, we have been working under the assumption that we'll be
+inserting version info into `traitId`, which in the current environment
+essentially makes each new version of a trait a unique trait, which has
+both pros and cons.
+
+A decision intentionally not made in this DR is whether or not this sort
+of presumption is the way forward with an in-data approach. it is
+possible to pack version data into `traitID` and yet not presume that
+`mediacreation-v1-locatableContent` is a conceptually different trait
+from `mediacreation-v2-locatableContent`.
+
+There are reasons we might want to do this, for example `entityTraits`
+and perhaps even `managementPolicy` could potentially benefit from being
+unversioned. Does a user want to know that an entity supports
+`locatableContent-v2` specifically, or are we simply asking "What is
+this thing?", in which case, an unversioned, or "all-version" method of
```

Comment:
I think there are some really interesting ideas there - you could even say that `resolve` with no version means “whatever you have”, I think `entityTraits` for read would be useful with versioning info as then you know what data you might be able to get (which helps significantly with migrations or user-facing comparability feedback, and writing generic browsers/views)

Reply:
We benefit from having generated trait views here, as it gives us options. I can imagine an "all-versions" view which presents the interface of every property that has ever existed in the history of the trait, but errors/excepts with version errors if the data dosen't line up. We'd have to think what that means re: `isImbued` etc, but the compatibility is appealing, and more than that, the idea that we can segregate versions away from identity is appealing. 

I can't help but thinkg that it's not gonna be a fun time if you're trying to find out if an entity is an image, but you get `false` results because _actually_ your entity is `imagev4` and you asked about `imagev3`, (because your build step hadn't ran traitgen recently enough, or whatever). I also worry about the difficulty of implementation on the managers side for version aware introspection methods, could get hairy.

I fear making this decision is going to take months and months...

Reply:
> get false results because actually your entity is imagev4 and you asked about imagev3

If we're saying Image v4 is a breaking change from Image v3 then can you definitely say they are equivalent and the false is probably correct? Or if they are suitably equivalent - did the traits need a new version?   Did I hear OTIO don't bump versions for additions? As we already have too manage missing properties where data is consumed.

Reply:
> I fear making this decision is going to take months and months...

Being cheeky: pick something that already exists instead of making a new thing, OTIO/USD/files all seem to be going in the same way, and that way also means its not a core API concern.

Reply:
Another potential benefit of aligning with in-data and OTIO or USD as closely as possible is that it’s not another thing for people to learn. 

If I understand correctly, it looks like it would also be a lot more straight forward to store/recall data from one in the other. Eg OTIO meta data or USD prim schemas could all become traits and visa versa.

Reply:
> Being cheeky: pick something that already exists instead of making a new thing

We're not trying to invent something new, really, it's just that OpenAssetIO has uniquenesses about it. We tried to just do what seemed simplest and most normal to us in the last DR to poor results...
The in-data approach here, which is what i'm leaning towards currently, _is_ a thing that already exists right, we've seen it around. It's mainly just its interaction with specifications that gives me pause.

> At the risk of being glib, I bet you could probably borrow the entire OTIO upgrade/downgrade mechanism and point it at at TraitsData with minor changes

As far as I recall, OTIO has a central registry of all possible types that can exist, and has strict requirements about the existence of upgrade/downgrade conversion methods. Ie, if you change the schema, they enforce the requirement for upgrade/downgrade methods to be committed along with the schema change, so it's not 1:1. The main thing is that we don't have visibility or control of external trail libraries that may appear, so we can't make strict guarantee about version upgrades occurring.

Nonetheless, this dosen't preclude in-data upgrade abilities in the trait libraries (ie mediacreation,) and we definately want to add those, but we're trying to limit the scope at first to a mechanism that can work without complex automated utilities, (this is based on TSC feedback, that for the moment, managers at least would be happy to do the conversions themselves. ... Hosts are perhaps another story, we wern't thinking about hosts doing conversions at that point.)

A thing to note is that writing those methods is more burdensome on contributors/us that it is on OTIO, because it must happen at the trait library level, we'd have to support at least 2 languages, ie 2 versions of each upgrade/downgrade function.

Also, the question of speed of iteration came up in regards to this, do we really want to enforce the need for upgrade/downgrade functions to be written when we know mediacreation is early days and requires iteration?

Reply:
> > Being cheeky: pick something that already exists instead of making a new thing
> 
> We're not trying to invent something new, really, it's just that OpenAssetIO has uniquenesses about it. 

I guess that was the cheeky part - is it _really_ that different if you step back far enough? I just has extras that can stray into think that it is different? An active, out-of-data version negotiation mechanism just felt relatively "new" in comparison to the prior art in the industry doing on the whole, similar things?

> The in-data approach here, which is what i'm leaning towards currently, _is_ a thing that already exists right, we've seen it around. It's mainly just its interaction with specifications that gives me pause.

Sure, personally I think the reality of specifications is maybe just that they're closer to what you've always talked about than perhaps they are being portrayed. Keeping them a 'programmer convenience managed by traitgen' feels to work for me, and then keeps them fairly compatible. Forcing the conversation about trait sets between independent parties. And you can then use specs as a bookmark for formally defining which one you need. Having a runtime registry of classes to trait sets is pretty straight forward (we had that before). Hosts/Managers shouldn't get too hung up on 1:1 specification matches, as that defeats the composition principal.

> As far as I recall, OTIO has a central registry of all possible types that can exist, and has strict requirements about the existence of upgrade/downgrade conversion methods. Ie, if you change the schema, they enforce the requirement for upgrade/downgrade methods to be committed along with the schema change, so it's not 1:1. The main thing is that we don't have visibility or control of external trail libraries that may appear, so we can't make strict guarantee about version upgrades occurring.

When I meant glibly "take OTIOs mechanism" - I meant take _all_ of it 🙃.
 
> Nonetheless, this dosen't preclude in-data upgrade abilities in the trait libraries (ie mediacreation,) and we definately want to add those, but we're trying to limit the scope at first to a mechanism that can work without complex automated utilities, (this is based on TSC feedback, that for the moment, managers at least would be happy to do the conversions themselves. ... Hosts are perhaps another story, we wern't thinking about hosts doing conversions at that point.)

I guess it felt like it all works without them, just you don't get stuff for free without code that makes it free? It's purely a choice that we make as end users as to how compatible/rigourous you want to be with inter for versions. Some hosts/managers may refuse to support anything other than the latest etc... regardless of what mechanisms you put in place - simply to avoid being in a world of infinite permutations to maintain.

> A thing to note is that writing those methods is more burdensome on contributors/us that it is on OTIO, because it must happen at the trait library level, we'd have to support at least 2 languages, ie 2 versions of each upgrade/downgrade function.

For sure - I wonder would it be any more effort than the whole mechanics and ongoing maintenance in every implementation of new core api methods and/or the negotiation handshake?

> Also, the question of speed of iteration came up in regards to this, do we really want to enforce the need for upgrade/downgrade functions to be written when we know mediacreation is early days and requires iteration?

I guess thats a choice of when you decide to change the version number, rather than the mechanism? Like how we treat alphas and betas differently to post v1 server?

Reply:
> but we're trying to limit the scope at first to a mechanism that can work without complex automated utilities, (this is based on TSC feedback, that for the moment, managers at least would be happy to do the conversions themselves. 

Thinking further on this - In a way - is that a benefit of the in-data approach? It can be explored with, I believe, no changes to the core API (other than potentially documentation), and manageable iterations to traitgen and select implementations that wish to participate.

With the in-data approach, If managers wish to do the work in the first instance, a host could leave its code essentially unchanged (bar picking up a new package version and perhaps minor class name changes), the manager can use version introspection mechanisms in traitgen code to see which versions a host is using and adapt accordingly if required - the manual work suggested. 

This would potentially only need some changes to trait-get and how the yaml is defined. Managers could even work with existing un-modified hosts treating them if traitgen treated them as "v0" or some such.

With the out-of-band approach, it looked like both parties would have to do some work to add in negotiation, and you'd need core API changes to introduce the mechanism.

It seems like either way, in the long run if host or manager code needs to "consider" versions, and there doesn't seem to be substantially different orders of magnitude of work between either solution, but the impact of change is significantly different.

There is one other perspective that, what ever the initial outcome of this decision process is, it will be considered lacking from _some_ world view or another and need refinement. The in-data solution seems to have a much smaller blast-radius than an actively negotiated mechanism that touches the core API. This would allow it to be more easily iterated through practical application in a much smaller part of the ecosystem, and also be much more amenable to future changes.

To me facilitating iteration in a more decoupled, less intrusive way is quite a significant benefit, given the uncertainly about the decision to date.

Reply:
Tbh i think that's captured in the "seems to fit with OpenAssetIO design" pro. Cause it does seem to fit.